### ISSUE TITLE
Support string typed metrics

### ISSUE THREAD

**Issue by andy-maier** [prometheus/prometheus#9139]
## Proposal
**Use case. Why is this important?**

The are cases where the interpretation of other metric values depends on mode-like properties. For example, on the Z system machines, logical partitions can dedicate or share the processors. That is represented as a string enum typed resource property "processor-mode" with values "shared" or "dedicated". That mode is important to know if you want to calculate derived metrics or define alerts based on other integer or float-typed metrics around the processors (I don't go into the details here). In our case, the processor mode of a logical partition can be changed over time, so it is appropriate to transfer it along with the other metrics whose interpretation depends on it.

Another use case is status values that someone wants to put into the metric store for correlating other metrics with it, or simply for long term storage of the status.

I did read issue #2227, but this is now 5 years old and there has been development since then. Other monitoring solutions meanwhile support string-typed metrics. Here are just some that I stumbled across:
* SysDig: https://docs.sysdig.com/en/metrics-dictionary.html.
* Nagios supports enumerated state values (https://assets.nagios.com/downloads/nagioscore/docs/nagioscore/3/en/statetypes.html).
* Zabbix supports items that can have "short character data" as a value (https://www.zabbix.com/documentation/current/manual/config/items/item).

**Alternatives**

I do understand that we can map string enum typed metrics into integer values, and that is also how we will start out supporting this in our exporter (https://github.com/zhmcclient/zhmc-prometheus-exporter), but it is just way more natural to be able to use the string enum values directly.

Another dimension of alternatives would be to transport string typed values from the monitored system using another monitoring solution that supports this, but that puts the burden on the user to now maintain two metric gatherer environments on each system, and it is also a slippery slope since it opens the door to future consolidation in the "wrong" direction.

**Comment by roidelapluie** [comment]
Thanks for this. I have been speaking with other people and it sounds like it would open the door for event based monitoring, if you expose strings with timestamps, which prometheus is not. If it is an enum, you could also map the state to particular integers.

For a more in-deep decision, I will link this issue in our next developers summit talks (where anyone can participate) to have broader discussions.

**Comment by RichiH** [comment]
I am aware that the below are not direct replies, but they still might be useful:
* OpenMetrics talks a bit about how to do ENUMs in Prometheus: https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#stateset
* Grafana Loki can store arbitrary strings and blobs, which can be queried with a language based on PromQL; and turn those queries into metrics

**Comment by rudigerlove** [comment]
any hopes in 2024 ? this is a very useful feature.

**Comment by beorn7** [comment]
For completeness, here is an excerpt from the meeting notes of the dev-summit where we discussed this issue (on 2021-11-18):

> Julien: Metrics values as string
https://github.com/prometheus/prometheus/issues/9139
> I do not think we should do it but let’s discuss broadly and openly.
> Beorn: I caught myself thinking about the “opposite” more often, recently: numbers as label values. (I think we should probably do neither at the moment, but it might be good to discuss both together.)
> 
> __CONSENSUS: We can see the use cases, but on balance do not think the use cases are worth the pain to support string-type metrics or numeric label values. This consensus is not designed to close the discussion, it’s just a point in time statement.__

**Comment by bboreham** [comment]
Hello from the bug-scrub!

While this is a valid feature request, as far as we know nobody has worked on it, and it is a large (many months) project.
Most of the points cited at https://github.com/prometheus/prometheus/issues/12233#issuecomment-2025848266 are relevant here.

If someone is interested to work on it, the first step would be to post a design document.

**Comment by beorn7** [comment]
There is already the concept of an info series (see OpenMetrics info metric and the experimental `info` function in PromQL).

Also, there is this idea of a meta-data store based on "meta-data series": https://docs.google.com/document/d/1epBslSSwRO2do4armx40fruStJy_PS6thROnPeDifz8/edit?usp=sharing

**Comment by ringerc** [comment]
[Related CNCF Slack thread](https://cloud-native.slack.com/archives/C01LSCJBXDZ/p1747783784617299)

**Comment by ringerc** [comment]
@beorn7 Info-metric series with changing labels have some really heavy impacts on the TSDB indexes and on memory use in queries, as noted above. This can rapidly become prohibitive, requiring the dedication of enormous amounts of memory to Prometheus to keep it stable.

Long-valued labels also use outsized resources.

So while it's _possible_ to just (ab)use an info-metric to carry a string-valued label, in practice this is unusable at any reasonable scale unless memory is free and infinite.

**Comment by beorn7** [comment]
To be clear: Info metrics are currently not supported natively. They are just a concept modeled on the classic Prometheus data types. The mentioned `info` function is a first step to support info metrics better – it would in fact need new concepts in the TSDB to be fully implemented. In that process, storing info metrics can also be improved. Which is closely related to the proposed dedicated metadata store.

**Comment by rdiez** [comment]
I wanted to create an GitHub issue titled "String data type for metrics" myself, but then I found this one. This is my point of view:

I am using Prometheus to monitor my systems, but not everything I want to monitor and query are numerical values:

- I to check whether MAC or IP addresses are duplicated in my VMs.
- I want to check if some serial numbers or disk batch numbers 2025-A3-Xnnn are known to have issues.
- I want to know which Linux machines are running Fedora vs Ubuntu.
- I want to check whether the IP addresses reported by my main and backup DynDNS providers are the same.
- I want to store a history of the IP addresses my Internet provider has assigned me, and warn me when it changes.
- I want to collect installed software versions and warn if LibreOffice is too old or hasn't been updated for a long time.
- I want to collect the E-Mail addresses of the sysadmins responsible for each computer (which are stored in a file in each machine), and warn me if it changes on a few selected servers which are important to me.
- I want to know how many of my Linux kernel versions are still older than 6.8.3.
A version from /proc/version_signature looks like: Ubuntu 6.8.0-65.68~22.04.1-generic 6.8.12
- I want to know which kernel versions a particular server has gone through in the last 3 years, in case a particular bug apparently only happens with one version.
- I want to know how often my Linux servers upgrade kernels, between major and/or minor versions. And which systems have not changed kernel versions for 3 weeks.
Etc. etc.

The trouble is, Prometheus metrics can only store numeric values. This is of course not sufficient, so people have used workarounds with labels. Every time I need a string, I hear the advice "add the string as a label, then set the value to a fixed 1".

One such workaround has been implemented in the Blackbox exporter for DNS probes:

probe_ip_addr_hash: Specifies the hash of the IP address. It’s useful to detect if the IP address changes.

Unfortunately, the hash is about the IP address of the DNS server used, not of the IP address returned by the DNS probe, but that is a shortcoming in the Blackbox exporter. While a hash is a good trick for detecting changes, it is not 100 % reliable (possible collisions), and you may want to see the original IP address, not its hash.

I am no Prometheus expert, but I guess the workaround of setting string values in labels has important drawbacks: probably worse query performance, unnecessary increase in the number of metrics (cardinality), and a broken long-time history for a particular metric instance.

Another workaround I hear often is: implement your own exporter, or use the textfile collector, and in your script, parse the string and generate a numeric value like 1 for good and 0 for bad. But then I have to distribute and upgrade such collector scripts everywhere. Some fixed enum cases can be reduced to numbers, but that has drawbacks too. What happens if a new, unforeseen string enum value suddenly comes up?

If Prometheus supported a string data type on metrics, I think it would make the checks and alerts I listed above much easier to implement. It would open new possibilities with little cost for the user.

When I searched the Internet about a string data type for Prometheus metrics, I did not find much, but I did see some unclear performance worries storing strings.

If you are really worried about performance, you could configure a fixed string size for such metrics (perhaps configurable per metric). After all, most strings I want to monitor do not change often and are not very long. You could also not collect any scraped strings by default, so that the user would have to specify what string metrics to collect per target. You could also document that string metrics have a lower performance, so the user should be frugal with them.

InfluxDB has supported the string data type for a long time:
https://docs.influxdata.com/influxdb/v2/reference/syntax/line-protocol/#string
https://docs.influxdata.com/influxdb3/core/reference/line-protocol/#string
See for example this discussion:
https://www.reddit.com/r/grafana/comments/rrgalu/timeseries_database_with_string_data/

Grafana can even graph string enums:
https://github.com/grafana/grafana/issues/14222

**Comment by beorn7** [comment]
> If you are really worried about performance,

I don't think that this is the primary worry. (Did anyone ever say that?)

I would say the primary worry is that this touches almost all layers of the stack (see the "component/..." labels added to this issue). It will be a very invasive change requiring a lot of work. Beyond that, there is the more subtle aspect that the increased code complexity is a long-term maintenance burden. What we learned when adding native histograms as a new native metric value type beyond "just floats", which was the only real metric value type from an internal perspective (counters and gauges were just conventions), is that we bloated the code base with a lot of switch statements to handle the different types. That doesn't scale. We need a more robust code structure to support a multitude of metric value types first. See also #12233 and #16949 for similar issues.

**Comment by ringerc** [comment]
Could this be done by continuing to (ab)use the info-metric pattern, storing the string-value data in a label?

Then defend against the impact on the inverted TSDB indexes by supporting a non-inverted covering index like a B-tree or B+-tree keyed by the name of the string-valued label, the series's identifying labels, then the oldest timestamp at which a given string-value appeared. This might reduce the scope of the change across the codebase by concentrating most of the change in the TSDB query/index handling. That way the data model and most functions etc would not need major changes (though of course the benefits would be reduced too).

I made some notes here https://gist.github.com/ringerc/c030ee6010a592ebb146b03991f14f09  on how such an approach might be designed and implemented.

But when thinking through it I found some problems that convince me that, even if the TSDB storage issues are disregarded, there are too many other parts of Prometheus that cannot gracefully cope with series that have highly-varying, potentially long-valued labels.

Lets say that this approach would permit highly varying string values could be stored and fetched efficiently, without the consequences of high label churn, long values and high cardinality on the TSDB and its indexes.

Prometheus still has no way to represent the results efficiently, nor any way to return a large result set without excessive working memory use.

Even if fetching `foo{}[1h]` where `foo{}` has some long-valued varying label is made more efficient on disk and for index searches, prometheus is going to OOM and crash too easily when accumulating a result to send to the client. The result is first  going to be multipled in-memory into a wide set of sparse range-series `foo{valuelabel="a"}`, `foo{valuelabel="b"}` when fetching the value-labels. Then marshalled into a single response payload, which must also fit in memory at the same time as the whole exploded result-series, including all the duplicates of all the identifying label key/value pairs. I don't see it working well.

To make this work _well_:

* A means would be needed to represent a series in-memory with some of its labels pointing to some common-set of labels shared between many series, so that `foo{}[1h]` with a string-value-label that changes every sample would not have to duplicate `job`, `instance`, and every other identifying label for every sample in which any value changes.
* The query executor would really want to be able to split up execution into batches with results that are progressively integrated after unwanted labels are discarded, etc. That's not something PromQL's rather simple executor can do right now, it does not have a dynamic query planner like a relational database does.
* It would be desirable to be able to stream query results to the client in batched chunks, rather than accumulating them in memory and returning them as a single huge result message payload that can easily OOM Prometheus.

... and all of these would probably be really intrusive code changes.

But without those things, having efficient storage isn't going to help that much, Prometheus is just going to crash on nontrivial queries of the results.

At least some of these issues would continue to be a problem even if there was a true, native string-valued metric data type. In particular, excessive working memory consumption leading to crashes because Prometheus cannot batch query execution in chunks or stream the results.

**Comment by rdiez** [comment]
I am a (non-expert) user, so your description goes way over my head. But nevertheless, I wanted to say that the string data type is so important, that I think a first, non-optimal implementation would be worth it, if only to get other software parts rolling.

This statement got my attention though:

> prometheus is going to OOM and crash too easily

I would have hoped that prometheus keeps track of the response size, and if it grows over some limit, it would just return an error like "out of resources" or "memory quota exceeded for query", instead of crashing due to the OOM. Such a memory limit per query could or should be configurable.

If that is the case, then I wouldn't mind having some queries with the new string data type failing often due to a first, inefficient implementation. The most useful string data I can think of are things which do not change often, like a version number. Would that reduce the memory consumption? The first string implementation could then state that prometheus cannot properly or efficiently deal with strings which change often.

This bug report also talks about implementing non-cardinal labels, and I have created a separate feature request for it which states another reason why such labels are helpful for config/alert purposes: #17057

If strings were stored in such non-cardinal labels, would that reduce the excessive memory consumption you are worried about? Or does your "series's identifying labels" already imply that you would be implementing the new labels as a separate label type? If so, maybe it would be worth implementing other non-cardinal label types (not just for string data type, but for config/alert purposes) at the same time.

**Comment by ringerc** [comment]
> The most useful string data I can think of are things which do not change often, like a version number

These are reasonably handled by the info-metrics convention already.

The only reason to want string-valued metrics would be to handle frequently changing values, where the current approach of carrying the value as a label blows out Prometheus's TSDB inverted index sizes and its in-memory loading of series labels during querying.

As for non-cardinal labels, I would like that too (a lot) but I can see that their implementation is far from trivial given Prometheus's schemaless, largely metadata-free nature and the assumptions baked into its indexes and query language.

**Comment by rdiez** [comment]
> These are reasonably handled by the info-metrics convention already.

You may be right. I am no Prometheus expert, but perhaps you can tell me how to implement these from the list above:

* I want to check whether the IP addresses reported by my main and backup DynDNS providers are the same.

* I want to store a history of the IP addresses my Internet provider has assigned me, and warn me when it changes.

* I want to know how often my Linux servers upgrade kernels, between major and/or minor versions. And which systems have not changed kernel versions for 3 weeks.

**Comment by ringerc** [comment]
@rdiez 

> You may be right. I am no Prometheus expert, but perhaps you can tell me how to implement these from the list above:
> 
> * I want to check whether the IP addresses reported by my main and backup DynDNS providers are the same.

Assuming you don't have huge numbers of rapidly changing dynamic DNS registrations this would be handled by (ab)using labels, carrying a non-identifying string in a label value.


```
# at t=0
my_current_dyndns_ip{host="foo",current_address="1.2.3.4"} 1

# at t=1
my_current_dyndns_ip{host="foo",current_address="9.9.9.9"} 1
```

Combined with appropriate use of `last_over_time(my_current_dyndns_ip[...])` over a small range interval, this helps reduce duplication.

Unfortunately Prometheus does not to my knowledge have a `latest` aggregation operation, which would make it much easier to query this data and enumerated stateful sets by picking the series with the most recent sample timestamp. I just raised a ticket suggesting that function as it's something I have wanted regularly: https://github.com/prometheus/prometheus/issues/17129

> * I want to store a history of the IP addresses my Internet provider has assigned me, and warn me when it changes.

Same as above.

> * I want to know how often my Linux servers upgrade kernels, between major and/or minor versions. And which systems have not changed kernel versions for 3 weeks.

This is a typical usage pattern for info-metrics already, see e.g. how `kube-state-metrics` and `node-exporter` expose versions as labels in constant-valued metrics series.

It's wasteful to store the constantly-repeating `1` values, but Prometheus is so astonishingly efficient at sample storage that it's just not that bad.

**Comment by bwplotka** [comment]
We started to discuss native stateset in Prometheus: https://github.com/prometheus/prometheus/issues/17914

Feel free to join this discussion!

AFAIK `StateSet` type satisfies most use cases mentioned for `string typed metric`. Is that true? Are folks here agreeing that stateset is the good enough answer to this problem? If yes, we can close this issue and address this problem in https://github.com/prometheus/prometheus/issues/17914

Notably [OM 1.0 StateSet does not allow any numbers per ](https://github.com/prometheus/OpenMetrics/blob/main/specification/OpenMetrics.md#stateset) "state" - only 0 and 1. 

Some use cases like `I want to store a history of the IP addresses my Internet provider has assigned me, and warn me when it changes.` and others, literally ask for higher cardinality acceptance. Adding a `map` type is some way of doing this, but likely out of scope for statset.

**Comment by rdiez** [comment]
> Some use cases like I want to store a history of the IP addresses my Internet provider has assigned me, and warn me when it changes. and others, literally ask for higher cardinality acceptance.

I am no Prometheus expert, but I gather the term "cardinality" in this context usually means the number of metrics due to different labels. I think that storing a string value is not related to this "cardinality" concept, as a string would virtually mean an infinite cardinality (expect for the restriction on the maximum string length).

I think that no enum type could really match the versatility of a string in many useful scenarios.

**Comment by ringerc** [comment]
@bwplotka IMO there are two related use cases here. One would be satisfied by a native state-set, one would not.

**TL;DR**: state-sets can be modelled with string-valued metrics, but there are many string-valued time-series that are not state-sets (e.g. IPv6 addresses), so state-sets cover some use cases for string-valued series but not others.

--

A native state-set would work well where the string-values represent values in an enumeration whose possible values are reasonably  limited in number, reasonably stable over time, and reasonably short. Some of the use cases discussed here fit this need. These are currently handled in Prometheus with a pretty ugly and inefficient model where 'n' samples are emitted for each 'n' possible states; active states have value 1, inactive states have value 0. This mostly works ok, but is very storage and memory inefficient, has issues with consistency, etc, as discussed in https://github.com/prometheus/prometheus/issues/17914 . An enumerated state-set is IMO very clearly a reasonable metrics sample, and OpenMetrics agrees.

The other use case is for non-state-set data that cannot be sensibly represented as a 64-bit integer but that [do not logically form a cardinal dimension to identify a new series](https://github.com/prometheus/prometheus/issues/11132). Examples include IPv6 addresses, UUIDs, SHA128/256/512 hashes, etc,is where the series represents "this is the current address" or similar. More generally, string-valued samples. While these can be kludged using labels, there are endless problems with reliably and consistently querying them, and they tend to cause grossly excessive disk and memory use in Prometheus. If sample values are highly variable over time it rapidly explodes the TSDBs inverted indexes, and overwhelms Prometheus's in-memory, fully-materializing executor. These are not enumerated state-sets, and cannot be sensibly represented as one. [See e.g. @rdiez 's mention of IPv6 addresses as an issue](https://github.com/prometheus/prometheus/issues/17914#issuecomment-3919114434). I previously explored the idea of using non-cardinal labels to handle this need in https://github.com/prometheus/prometheus/issues/11132 but I got stuck trying to figure out how to fit it into the Prometheus metric model, executor and database format.

The current argument in https://github.com/prometheus/prometheus/issues/2227 seems to be that the latter sort of data is just not well suited for handling by Prometheus - Prometheus is the wrong tool for the job and doesn't want to compromise its core goals to become better at handling this sort of data. This kind of data isn't a metric, and Prometheus is a metrics database. I can see the counter-argument too; such data can still be seen as a valid _time series of states_, such as the current peer IP address of a connected service, it's just not _metrics_; it's reasonable to want to use a time-series DB to record and track it.
